File: 316-remove-duplicate-letters/remove-duplicate-letters.java
class Solution {
    public String removeDuplicateLetters(String s) {

        // Count frequency of every character
        int[] count = new int[26];

        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
        }

        // Stack to build the answer
        Deque<Character> stack = new ArrayDeque<>();

        // Track characters currently in the stack
        boolean[] used = new boolean[26];

        for (int i = 0; i < s.length(); i++) {

            char c = s.charAt(i);

            // This occurrence has now been processed
            count[c - 'a']--;

            // Already present in answer
            if (used[c - 'a']) {
                continue;
            }

            // Greedily remove larger characters
            // if they appear again later
            while (!stack.isEmpty()
                    && c < stack.peek()
                    && count[stack.peek() - 'a'] > 0) {

                char removed = stack.pop();
                used[removed - 'a'] = false;
            }

            // Add current character
            stack.push(c);
            used[c - 'a'] = true;
        }

        // Build answer in reverse
        StringBuilder ans = new StringBuilder();

        while (!stack.isEmpty()) {
            ans.append(stack.pop());
        }

        return ans.reverse().toString();
    }
}

File: 316-remove-duplicate-letters/remove-duplicate-letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        //here we use the stack 
        vector<int> chars(26,0);
        stack<char> st;
        for(int i = 0; i < s.size(); i++)
        {
            chars[s[i] - 'a']++;
        }
        //now we have to store the current stack 
        unordered_set<char> current;
        for(int i = 0; i < s.size(); i++)
        {
            if(current.count(s[i]))
            {
                chars[s[i] - 'a']--;
                continue;
            }
            while(!st.empty() && s[i] < st.top() && chars[st.top() - 'a'] > 1)
            {
                chars[st.top() - 'a']--;
                current.erase(st.top());
                st.pop();
            }
            st.push(s[i]);
            current.insert(s[i]);
        }
        string ans = "";
        while(!st.empty())
        {
            ans += st.top();
            st.pop();
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};